### PR TITLE
Unit tests

### PR DESCRIPTION
**Pull request by nunesjoab**
* **Please check if the PR fulfills these requirements**
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
This PR introduces a set of unit tests to iotagent-mosca

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Is there any issue related to this PR in other repository?** (such as dojot/dojot)
dojot/dojot/1187

* **Other information**:

### CODE REVIEW

File: coverage/clover.xml [hunk 1]
```diff
@@ -0,0 +1,408 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (mprevide) [inline coverage/clover.xml]: Files within coverage/ really need to be versioned?

**Comment by mprevide** [comment]
I added codecov to my PR draft (https://github.com/dojot/iotagent-mosca/pull/74/commits/ac9aab23f682c4437bb1955602e2aa8af4c868ac), so maybe it's a good idea to add in this PR too. And I associated the project with the codecov service.

File: coverage/clover.xml (hunk 1 above)
Comment (nunesjoab) [reply coverage/clover.xml]: I don't think so but I'm not sure if they are used by other tools like Travis or Codefactor. I will add the coverage path to gitignore. If necessary, I can remove this line later.

**Comment by nunesjoab** [comment]
> I added codecov to my PR draft ([ac9aab2](https://github.com/dojot/iotagent-mosca/commit/ac9aab23f682c4437bb1955602e2aa8af4c868ac)), so maybe it's a good idea to add in this PR. And I associated the project with the codecov service.

Added :)

**Comment by mprevide** [comment]
Maybe it's necessary remove coverage from gitignore and add something like below in package.json: 


"jest": {
    "coverageDirectory": "./coverage/",
    "collectCoverage": true,
    "collectCoverageFrom": [
      "lib/**/*.js"
    ]
  }

**Comment by mprevide** [comment]
I think you should actually remove /covere/ * from gitignore.

File: coverage/clover.xml (hunk 1 above)
Comment (giovannicuriel) [reply coverage/clover.xml]: After running the unit tests, codecov uploads the results to its site. There's no need track them in git.

File: src/iotagent.js [hunk 2]
```diff
@@ -128,7 +128,15 @@ class IoTAgent {
...
-    return topic.split('/')[index];
+    if(topic) {
+      if(topic.includes('/')) {
+        return topic.split('/')[index];
+      } else {
+        return topic;
+      }
+    }
+
+    return undefined;
```
Comment (giovannicuriel) [inline src/iotagent.js]: You can also just `return`.

File: tests/unit-tests/iotagent.test.js [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+
+        const event = {
+            data: {
+                id: 235,
+            }
+        };
+
+        expect(iotagent._processDeviceRemoval('admin', event)).toBeUndefined();
+    });
+
```
Comment (jonaphael) [inline tests/unit-tests/iotagent.test.js]: to increase the add this case to test device actuation function

` // test a device actuation
 it("should process a device actuation ", () => {

        const tenant = 'my-tenant';

        const event = {
            data: {
                id: 1,
                attrs: {
                    'attr-1': 'attr-1-value'
                }
            }
        };

        const deviceID = event.data.id;
        const topic = `/${tenant}/${deviceID}/config`;

        // messages builded
        const message = {
            'topic': topic,
            'payload': JSON.stringify(event.data.attrs),
            'qos': 0,
            'retain': false
          };

          // >> Tested code
        const iotagent = new IoTAgent();
        iotagent.mqttBackend.server.publish = jest.fn();
        iotagent._processDeviceActuation(tenant, event);
        // << Tested code

        const firstArgument = iotagent.mqttBackend.server.publish.mock.calls[0][0];

        // >> Result verification
        expect(firstArgument).toEqual(message);
        // << Result verification
    })`

File: tests/unit-tests/metrics.test.js [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        const metrics = new moscaMestrics.Metrics();
+        metrics.preparePayloadObject()
+        expect(metrics.lastMetricsInfo.connectedClients).toBeNull();
+
+        metrics.preparePayloadObject(0);
+        expect(metrics.lastMetricsInfo.connectedClients).toBeNull();
+
+        const metricsAttribute = "connectedClients";
+        metrics.preparePayloadObject(metricsAttribute, 20);
+        expect(metrics.lastMetricsInfo[`${metricsAttribute}`]).toEqual("20");
```
Comment (jonaphael) [inline tests/unit-tests/metrics.test.js]: this should be another test case because this case expects null.

File: tests/unit-tests/metrics.test.js [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+            "connectionsLoad15min": 0,
+            "messagesLoad1min": 0,
+            "messagesLoad5min": 0,
+            "messagesLoad15min": 0
+        });
+
+        metricStore.lastMetricsInfo = null;
+        app.use(moscaMestrics.getHTTPRouter(metricStore));
+        response = await get('/iotagent-mqtt/metrics', '');
+        expect(response.status).toEqual(500);
```
Comment (jonaphael) [inline tests/unit-tests/metrics.test.js]: separate 200 test case from 500

File: tests/unit-tests/iotagent.test.js [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+        const serverArgs = mosca.Server.mock.calls[0][0];
+        const mqttInterfaces = serverArgs.interfaces;
+        expect(mqttInterfaces.length).toEqual(1);
+        expect(mqttInterfaces[0].type).toEqual("mqtts");
+        expect(mqttInterfaces[0].port).toEqual(8883);
+    });
+
+    it("Should build a MqttBackend instance with support for unsecure connections", () => {
+        // const config = JSON.parse(JSON.stringify(defaultConfig));
```
[the remark this answers is not in the document]
Comment (nunesjoab) [reply tests/unit-tests/iotagent.test.js]: Removed

File: tests/unit-tests/iotagent.test.js [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+        iotagent.agent = new Agent();
+        iotagent.mqttBackend = new backend.MqttBackend(iotagent.agent);
+        iotagent.metricsStore = new Metrics();
+
+        const samplePacket = {
+            topic: "$SYS/Yoda/",
+            payload: { toString: () => "sample-payload" }
+        }
+
+        expect(iotagent).toBeDefined();
```
[the remark this answers is not in the document]
Comment (nunesjoab) [reply tests/unit-tests/iotagent.test.js]: removed

File: tests/unit-tests/iotagent.test.js [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+        expect(iotagent.mqttBackend).toBeDefined();
```
[the remark this answers is not in the document]
Comment (nunesjoab) [reply tests/unit-tests/iotagent.test.js]: Removed

File: tests/unit-tests/iotagent.test.js [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+    it("Should process internal messages properly", () => {
+        const iotagent = new IoTAgent();
+        const healthcheck = new HealthChecker();
+        iotagent.agent = new Agent();
+        iotagent.mqttBackend = new backend.MqttBackend(iotagent.agent);
+        iotagent.metricsStore = new Metrics();
+
+        iotagent.initHealthCheck(healthcheck.healthChecker);
+        expect(iotagent.messageMonitor.monitor).toBeDefined();
+        // expect(iotagent._registerCallbacks()).toBeTruthy();
```
[the remark this answers is not in the document]
Comment (nunesjoab) [reply tests/unit-tests/iotagent.test.js]: Removed

File: tests/unit-tests/iotagent.test.js [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+
+/**
+ * Unit tests for iotagent-mosca
+ *
+ * This module has the following dependencies:
+ *
+ * - mosca
+ */
+
+// var Agent = require('@dojot/iotagent-nodejs').IoTAgent;
```
[the remark this answers is not in the document]
Comment (nunesjoab) [reply tests/unit-tests/iotagent.test.js]: Removed

File: tests/unit-tests/iotagent.test.js (hunk 3 above)
Comment (nunesjoab) [reply tests/unit-tests/iotagent.test.js]: Added

File: tests/unit-tests/iotagent.test.js [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+        const samplePacket = {
+            topic: "$SYS/Yoda/",
+            payload: { toString: () => "sample-payload" }
+        }
+
+        expect(iotagent).toBeDefined();
+        expect(iotagent.mqttBackend).toBeDefined();
+        expect(iotagent.getTopicParameter(samplePacket.topic, 1)).toEqual('Yoda');
+        expect(iotagent.getTopicParameter()).toBeUndefined();
+        expect(iotagent.getTopicParameter('topic')).toEqual('topic');
```
[the remark this answers is not in the document]
Comment (nunesjoab) [reply tests/unit-tests/iotagent.test.js]: Separated to another It

File: tests/unit-tests/metrics.test.js (hunk 4 above)
Comment (nunesjoab) [reply tests/unit-tests/metrics.test.js]: Separated to another it

File: tests/unit-tests/metrics.test.js (hunk 5 above)
Comment (nunesjoab) [reply tests/unit-tests/metrics.test.js]: test of status 500 removed because stopApp can't be called to change value of isInitialized variable.
